File: problems/12_advanced_graphs/0269_alien_dictionary/solution.py
"""
Problem: 0269. Alien Dictionary
Difficulty: Hard
URL: https://leetcode.com/problems/alien-dictionary/

解法：建圖 + 拓撲排序。
"""
from __future__ import annotations
from typing import List, Dict, Set
from collections import deque


class Solution:
    def alienOrder(self, words: List[str]) -> str:
        graph: Dict[str, Set[str]] = {c: set() for w in words for c in w}
        indeg: Dict[str, int] = {c: 0 for c in graph}
        for i in range(len(words) - 1):
            a, b = words[i], words[i + 1]
            ml = min(len(a), len(b))
            if len(a) > len(b) and a[:ml] == b[:ml]:
                return ""
            for j in range(ml):
                if a[j] != b[j]:
                    if b[j] not in graph[a[j]]:
                        graph[a[j]].add(b[j])
                        indeg[b[j]] += 1
                    break
        q = deque(c for c in indeg if indeg[c] == 0)
        out: List[str] = []
        while q:
            c = q.popleft()
            out.append(c)
            for nb in graph[c]:
                indeg[nb] -= 1
                if indeg[nb] == 0:
                    q.append(nb)
        if len(out) != len(indeg):
            return ""
        return "".join(out)


def _is_valid(order: str, words: List[str]) -> bool:
    seen = set("".join(words))
    if set(order) != seen:
        return False
    pos = {c: i for i, c in enumerate(order)}
    for i in range(len(words) - 1):
        a, b = words[i], words[i + 1]
        ml = min(len(a), len(b))
        diff = -1
        for j in range(ml):
            if a[j] != b[j]:
                diff = j; break
        if diff == -1:
            if len(a) > len(b):
                return False
            continue
        if pos[a[diff]] >= pos[b[diff]]:
            return False
    return True


if __name__ == "__main__":
    sol = Solution()
    o1 = sol.alienOrder(["wrt", "wrf", "er", "ett", "rftt"])
    assert _is_valid(o1, ["wrt", "wrf", "er", "ett", "rftt"])
    assert sol.alienOrder(["z", "x"]) in ("zx",)
    assert sol.alienOrder(["z", "x", "z"]) == ""
    assert sol.alienOrder(["abc", "ab"]) == ""
    o5 = sol.alienOrder(["a"])
    assert o5 == "a"
    o6 = sol.alienOrder(["ab", "adc"])
    assert _is_valid(o6, ["ab", "adc"])
    print("All tests passed!")

File: problems/12_advanced_graphs/0269_alien_dictionary/solution.swift
// Problem: 0269. Alien Dictionary
// Difficulty: Hard

import Foundation

class Solution {
    func alienOrder(_ words: [String]) -> String {
        var graph: [Character: Set<Character>] = [:]
        var indeg: [Character: Int] = [:]
        for w in words {
            for c in w {
                if graph[c] == nil { graph[c] = []; indeg[c] = 0 }
            }
        }
        for i in 0..<(words.count - 1) {
            let a = Array(words[i]), b = Array(words[i + 1])
            let ml = min(a.count, b.count)
            if a.count > b.count && Array(a.prefix(ml)) == Array(b.prefix(ml)) {
                return ""
            }
            for j in 0..<ml {
                if a[j] != b[j] {
                    if !graph[a[j]]!.contains(b[j]) {
                        graph[a[j]]!.insert(b[j])
                        indeg[b[j]]! += 1
                    }
                    break
                }
            }
        }
        var q: [Character] = []
        for (k, v) in indeg where v == 0 { q.append(k) }
        var head = 0
        var out = ""
        while head < q.count {
            let c = q[head]; head += 1
            out.append(c)
            for nb in graph[c]! {
                indeg[nb]! -= 1
                if indeg[nb]! == 0 { q.append(nb) }
            }
        }
        return out.count == indeg.count ? out : ""
    }
}

func validAlienOrder(_ order: String, _ words: [String]) -> Bool {
    var seen: Set<Character> = []
    for w in words { for c in w { seen.insert(c) } }
    if Set(order) != seen { return false }
    var pos: [Character: Int] = [:]
    for (i, c) in order.enumerated() { pos[c] = i }
    for i in 0..<(words.count - 1) {
        let a = Array(words[i]), b = Array(words[i + 1])
        let ml = min(a.count, b.count)
        var diff = -1
        for j in 0..<ml { if a[j] != b[j] { diff = j; break } }
        if diff == -1 {
            if a.count > b.count { return false }
            continue
        }
        if (pos[a[diff]] ?? -1) >= (pos[b[diff]] ?? -1) { return false }
    }
    return true
}

let sol = Solution()
let o1 = sol.alienOrder(["wrt", "wrf", "er", "ett", "rftt"])
if !validAlienOrder(o1, ["wrt", "wrf", "er", "ett", "rftt"]) { fatalError("a") }
if sol.alienOrder(["z", "x"]) != "zx" { fatalError("b") }
if sol.alienOrder(["z", "x", "z"]) != "" { fatalError("c") }
if sol.alienOrder(["abc", "ab"]) != "" { fatalError("d") }
if sol.alienOrder(["a"]) != "a" { fatalError("e") }
let o6 = sol.alienOrder(["ab", "adc"])
if !validAlienOrder(o6, ["ab", "adc"]) { fatalError("f") }
print("All tests passed!")
